File: packages/core/src/murmur.ts
/**
 * MurmurHash3 x86_32. Reference implementation for every Flagship SDK.
 *
 * See spec/BUCKETING.md. The traps this implementation exists to avoid:
 *
 * - JavaScript numbers are doubles, and `*` on two large uint32s loses low bits
 *   to float rounding. Every 32-bit multiply goes through Math.imul.
 * - `<<` and `|` yield *signed* int32, so every step is forced back to unsigned
 *   with `>>> 0`. Skipping this returns negative values that shift the bucket.
 */

const C1 = 0xcc9e2d51;
const C2 = 0x1b873593;

function rotl32(x: number, r: number): number {
  return ((x << r) | (x >>> (32 - r))) >>> 0;
}

/** Hashes raw bytes. Callers hash the UTF-8 encoding of a string, never UTF-16 code units. */
export function murmurHash3x86_32(data: Uint8Array, seed = 0): number {
  const len = data.length;
  const nblocks = len >>> 2;
  let h1 = seed >>> 0;

  for (let i = 0; i < nblocks; i++) {
    const o = i * 4;
    // Little-endian block read.
    let k1 =
      (data[o]! | (data[o + 1]! << 8) | (data[o + 2]! << 16) | (data[o + 3]! << 24)) >>> 0;

    k1 = Math.imul(k1, C1) >>> 0;
    k1 = rotl32(k1, 15);
    k1 = Math.imul(k1, C2) >>> 0;

    h1 = (h1 ^ k1) >>> 0;
    h1 = rotl32(h1, 13);
    h1 = (Math.imul(h1, 5) + 0xe6546b64) >>> 0;
  }

  // Tail: 0-3 bytes that did not fill a block. The fallthrough is deliberate
  // and matches the reference implementation.
  const tail = nblocks * 4;
  let k1 = 0;
  switch (len & 3) {
    case 3:
      k1 ^= data[tail + 2]! << 16;
    // fallthrough
    case 2:
      k1 ^= data[tail + 1]! << 8;
    // fallthrough
    case 1:
      k1 ^= data[tail]!;
      k1 = Math.imul(k1 >>> 0, C1) >>> 0;
      k1 = rotl32(k1, 15);
      k1 = Math.imul(k1, C2) >>> 0;
      h1 = (h1 ^ k1) >>> 0;
  }

  // Finalization mix.
  h1 = (h1 ^ len) >>> 0;
  h1 = (h1 ^ (h1 >>> 16)) >>> 0;
  h1 = Math.imul(h1, 0x85ebca6b) >>> 0;
  h1 = (h1 ^ (h1 >>> 13)) >>> 0;
  h1 = Math.imul(h1, 0xc2b2ae35) >>> 0;
  h1 = (h1 ^ (h1 >>> 16)) >>> 0;

  return h1;
}

const encoder = new TextEncoder();

/** Convenience wrapper that hashes a string's UTF-8 bytes. */
export function murmurHash3x86_32String(input: string, seed = 0): number {
  return murmurHash3x86_32(encoder.encode(input), seed);
}

File: sdks/go/murmur.go
package flagship

import "math/bits"

// MurmurHash3 x86_32. Must agree byte-for-byte with the TypeScript and Java
// SDKs — see spec/BUCKETING.md.
//
// Go needs none of the defensive masking the JavaScript port requires: uint32
// arithmetic wraps natively, and a Go string is already UTF-8, so []byte(s)
// yields exactly the bytes the spec calls for.
const (
	c1 uint32 = 0xcc9e2d51
	c2 uint32 = 0x1b873593
)

func murmurHash3x86_32(data []byte, seed uint32) uint32 {
	h1 := seed
	nblocks := len(data) / 4

	for i := 0; i < nblocks; i++ {
		o := i * 4
		k1 := uint32(data[o]) |
			uint32(data[o+1])<<8 |
			uint32(data[o+2])<<16 |
			uint32(data[o+3])<<24

		k1 *= c1
		k1 = bits.RotateLeft32(k1, 15)
		k1 *= c2

		h1 ^= k1
		h1 = bits.RotateLeft32(h1, 13)
		h1 = h1*5 + 0xe6546b64
	}

	// Tail: 0-3 bytes that did not fill a block. Fallthrough is deliberate.
	tail := data[nblocks*4:]
	var k1 uint32
	switch len(tail) {
	case 3:
		k1 ^= uint32(tail[2]) << 16
		fallthrough
	case 2:
		k1 ^= uint32(tail[1]) << 8
		fallthrough
	case 1:
		k1 ^= uint32(tail[0])
		k1 *= c1
		k1 = bits.RotateLeft32(k1, 15)
		k1 *= c2
		h1 ^= k1
	}

	// Finalization mix.
	h1 ^= uint32(len(data))
	h1 ^= h1 >> 16
	h1 *= 0x85ebca6b
	h1 ^= h1 >> 13
	h1 *= 0xc2b2ae35
	h1 ^= h1 >> 16

	return h1
}

// MurmurHash3String hashes a string's UTF-8 bytes.
func MurmurHash3String(input string, seed uint32) uint32 {
	return murmurHash3x86_32([]byte(input), seed)
}
